packages/protocol/contracts/identity/FederatedAttestations.sol

```diff
@@ -0,0 +1,530 @@
+   * @dev Throws if an attestation with the same (identifier, issuer, account) already exists
+   */
+  function registerAttestationAsIssuer(
+    bytes32 identifier,
+    address issuer,
+    address account,
+    uint64 issuedOn
+  ) external {
+    // TODO allow for updating existing attestation by only updating signer and publishedOn
+    require(issuer == msg.sender);
+    _registerAttestation(identifier, issuer, account, issuer, issuedOn);
+  }
+
+  /**
+   * @notice Registers an attestation with a valid signature
+   * @param identifier Hash of the identifier to be attested
+   * @param issuer Address of the attestation issuer
+   * @param account Address of the account being mapped to the identifier
+   * @param issuedOn Time at which the issuer issued the attestation in Unix time 
+   * @param signer Address of the signer of the attestation
+   * @param v The recovery id of the incoming ECDSA signature
+   * @param r Output value r of the ECDSA signature
+   * @param s Output value s of the ECDSA signature
+   * @dev Throws if an attestation with the same (identifier, issuer, account) already exists
+   */
+  function registerAttestation(
+    bytes32 identifier,
+    address issuer,
+    address account,
+    address signer,
+    uint64 issuedOn,
+    uint8 v,
+    bytes32 r,
+    bytes32 s
+  ) external {
+    // TODO allow for updating existing attestation by only updating signer and publishedOn
+    validateAttestationSig(identifier, issuer, account, signer, issuedOn, v, r, s);
+    _registerAttestation(identifier, issuer, account, signer, issuedOn);
+  }
+
+  /**
+   * @notice Revokes an attestation 
+   * @param identifier Hash of the identifier to be revoked
+   * @param issuer Address of the attestation issuer
+   * @param account Address of the account mapped to the identifier
+   * @dev Throws if sender is not the issuer, signer, or account
+   */
+  function revokeAttestation(bytes32 identifier, address issuer, address account) external {
+    require(
+      account == msg.sender ||
+        // Minor gas optimization to prevent storage lookup in Accounts.sol if issuer == msg.sender
+        issuer == msg.sender ||
+        getAccounts().attestationSignerToAccount(msg.sender) == issuer,
+      "Sender does not have permission to revoke this attestation"
+    );
+    _revokeAttestation(identifier, issuer, account);
+  }
+
+  /**
+   * @notice Revokes attestations [identifiers <-> accounts] from issuer
+   * @param issuer Address of the issuer of all attestations to be revoked
+   * @param identifiers Hash of the identifiers
+   * @param accounts Addresses of the accounts mapped to the identifiers
+   *   at the same indices
+   * @dev Throws if the number of identifiers and accounts is not the same
+   * @dev Throws if sender is not the issuer or currently registered signer of issuer
+   * @dev Throws if an attestation is not found for identifiers[i] <-> accounts[i]
+   */
+  function batchRevokeAttestations(
```

Comment:
So for batch revoking there was this general conversation about passing in array indexes to avoid doing array lookups on-chain. This would be nice as an optimization but it would also require a significant rework for the batch because of the way array deletion happens via swap+pop. 
Essentially the only way to make it work is by ensuring that you're always removing elements from an array starting at the largest index.
```
list = [ a, b, c, d, e, f, g, h ]
items_to_remove = [ h, d, a ]
indexes_of_items_to_remove = [ 7, 3, 0 ]

// remove h - no swap
list = [ a, b, c, d , e, f, g ]

// remove d - swap with g
list = [ a, b, c, g, e, f ]

// remove a - swap with f
list = [f, b, c, g, e] 
```

If we were to remove the same items via swapping but out of order, for example if we started with `a`, we would actually swap that with `h` and our indexes would be ruined. from the get-go. 

So in order to make this operation work we would need to do something like:
```
function batchRevokeAttestations(
  address issuer,
  bytes32[] calldata identifiers,
  address[] calldata accounts,
  uint256[] identifierToAttestationsIndexOrdered,
  uint256[] addressToIdentifiersIndexOrdered
)
```

But the problem here is that `identifierToAttestationsIndexOrdered` sorting by this and sorting by `addressToIdentifiersIndexOrdered` might result in a different ordering for `identifiers` `accounts`.
Therefore we can't think of this as a tuple and can't construct is in such a way that there's a relationship between `identifier[i]`, `accounts[i]`, `identifierToAttestationsIndexOrdered[i]` and `addressToIdentifiersIndexOrdered[i]`... sadly, or at least I couldn't figure out if there's a property that would allow this to happen. If we only take into account array construction it would preserve ordering but because deletion causes weird swapping you can't really know. 

So what I would do is actually encode two `uint128` values in those `IndexOrdered` arrays, the first would be the actual array lookup index in the mapping and the 2nd would be the index in the `identifiers` and `accounts` array of the items that this lookup index refers to. Yeah I know... alternatively we would need to pass in 4 arrays and the stack wouldn't be happy. 

Ok once we have all of this we should be able to do one pass each of the `identifierToAttestationsIndexOrdered` and `addressToIdentifiersIndexOrdered` arrays and swap+pop each item from the mapping arrays, and during one of these passes also write to the `revokedAttestations` hash.

Again, this could be simpler if I could prove that the `attestationIndex` and `identifierIndex` in the mapping arrays maintain a consistent ordering so that two attestations would never end up with:
```
index_in_addressToIdentifiers_of(att1) > index_in_addressToIdentiifers_of(att2) 
and
index_in_identifierToAttestations_of(att1) < index_in_identifierToAttestations_of(att2)
```

Something like that. While writing this I got a glimmer of hope regarding a proof but I lost it. I'm gonna think about it some more tho. Also this might not be intelligible, which is fine 😂 I didn't do a great job of explaining it.

Reply:
> I didn't do a great job of explaining it.
false! this was a great explanation + thanks for meticulously documenting this issue 🙏 

(For now resolution for future readers) we went with limited arrays for the moment, reason being:
- the normal/expected use cases shouldn't get close to hitting that limit
- this allows us to make some guarantees about batch revocation without indices (and is much simpler than requiring indices with our swap/pop deletion)
- if many folks are using this (non-maliciously) and hitting the limit, we want to know + better understand those use cases since we probably need to rethink the solution that has those in mind 😄 

related PRs/links
- [batch revocation](https://github.com/celo-org/celo-monorepo/pull/9647)
- [array limits](https://github.com/celo-org/celo-monorepo/pull/9656)